ciao-controller/instance.go

```diff
@@ -285,6 +321,91 @@ func controllerStorageResourceFromPayload(volume payloads.StorageResource) (s ty
 	return
 }
 
+func networkConfig(ctl *controller, tenant *types.Tenant, networking *payloads.NetworkResources, cnci bool) error {
+	networking.VnicUUID = uuid.Generate().String()
+
+	if cnci {
+		hwaddr, err := newHardwareAddr()
+		if err != nil {
+			return err
+		}
+
+		networking.VnicMAC = hwaddr.String()
+		return nil
+	}
+
+	ipAddress, err := ctl.ds.AllocateTenantIP(tenant.ID)
+	if err != nil {
+		fmt.Println("Unable to allocate IP address: ", err)
+		return err
+	}
+
+	networking.VnicMAC = newTenantHardwareAddr(ipAddress).String()
+
+	// send in CIDR notation?
+	networking.PrivateIP = ipAddress.String()
+	mask := net.IPv4Mask(255, 255, 255, 0)
+	ipnet := net.IPNet{
+		IP:   ipAddress.Mask(mask),
+		Mask: mask,
+	}
+	networking.Subnet = ipnet.String()
+
+	cnciInstance, err := tenant.CNCIctrl.GetSubnetCNCI(networking.Subnet)
```

Comment:
the CNCI is created when we allocate the tenant IP address down in the datastore. It will know whether this is a new/different subnet at that time and do the launch if needed.